### ISSUE TITLE
seq.hotSPOT Bioconductor Submission

### ISSUE THREAD

**Issue by sydney-grant** [Bioconductor/Contributions#2922]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/sydney-grant/seq.hotSPOT

Confirm the following by editing each check box to '[x]'

- [x ] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x ] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x ] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x ] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x ] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x ] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x ] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x ] The 'devel' branch for new packages and features.
- [x ] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x ] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by vjcitn** [comment]
Please run R CMD check on your submission.  Note the License: field in DESCRIPTION.  Note also that the vignette has 
```
![NOTE](C:/Users/sy43548/OneDrive - Roswell Park Comprehensive Cancer Center/Documents/hotSPOT/vignettes/hotSPOT_ampfinder.jpeg)
```
Please improve the github repo README.md

**Comment by sydney-grant** [comment]
Hello,

The license in the description file has been updated. The file address in the vignette has also been updated. README.md file was changed to include package installation instructions.

Thanks,
Sydney
________________________________
From: Vince Carey ***@***.***>
Sent: Saturday, February 11, 2023 6:37 AM
To: Bioconductor/Contributions ***@***.***>
Cc: Grant, Sydney ***@***.***>; Mention ***@***.***>
Subject: EXTERNAL: Re: [Bioconductor/Contributions] seq.hotSPOT Bioconductor Submission (Issue #2922)


Please run R CMD check on your submission. Note the License: field in DESCRIPTION. Note also that the vignette has

![NOTE](C:/Users/sy43548/OneDrive - Roswell Park Comprehensive Cancer Center/Documents/hotSPOT/vignettes/hotSPOT_ampfinder.jpeg)


Please improve the github repo README.md

—
Reply to this email directly, view it on GitHub<https://secure-web.cisco.com/1Lz9C33VaELQUMYAQXB7okWluwRjfNRR4RM0wiEbstx1vicdsexF73dKZZ1q6W2VfuVXcDsqIBSKqjFzsJGcbcr8WAU-bQ23Oxnfvf4dYf3BHpK9z90W3gpDcPnraxqhIrpRw5ZRItLz8ZjP9wqX7TrhB3-BKsT82iH_xlj2c0pSzU8sWKkLD-XuOQP-uaiRotqN_Mr4TXJTGSoXyEsftRDxqqHpvzgPqWsYccpxMRPZPyuQzaeKe73AtJ3-Xtiqe6WO5Is8ewfZ_LMyNeLkcR4nKvpcYb_DFXGd7liK4BvWDNYleCk_ouGVs0SwZQg8oR-5NyP7wtLNy-_brQ4XV3w/https%3A%2F%2Fgithub.com%2FBioconductor%2FContributions%2Fissues%2F2922%23issuecomment-1426705949>, or unsubscribe<https://secure-web.cisco.com/1UBe1OuUjmAfnoCbWOTZj6u92ikolsDGNgeAZQOi2wpSDuovGrFa7b1z-HH_0mcqJfiVYjOFhbZyANmW-fKyEuK0_8TsLqtmCe5e81gQOOkXu3VnY_ptq_BGdn_pvicVl-inesttArJAXrjg4trCcb2Im1EWem5fB1nFn_w0hWfZDAHE69Jbx5J7DSt2siEvCBQwV8ZoIag4vw3Y20F9DK2z5A_VQIPKrxIcscf8WXQZiKGvwSvhdZYn9D8_3_3BqxBG8rF61xJbQN2iR-jM7JjQivm3ESS-F1gbk3Sa_6M1xF4YpcYbkbhM_4U-j1_QpUoVl-pQ97D3OoSpHn4tpvQ/https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAZUHEBORD2QIDC2BJNUERXLWW52XLANCNFSM6AAAAAAUVUFYNM>.
You are receiving this because you were mentioned.Message ID: ***@***.***>


This email message may contain legally privileged and/or confidential information.  If you are not the intended recipient(s), or the employee or agent responsible for the delivery of this message to the intended recipient(s), you are hereby notified that any disclosure, copying, distribution, or use of this email message is prohibited.  If you have received this message in error, please notify the sender immediately by e-mail and delete this email message from your computer. Thank you.

**Comment by vjcitn** [comment]
Please check your package on a clean system.  You have
```{r load library}
library(hotSPOT)
```
in the vignette and that package does not exist and is not defined in your submission.

**Comment by sydney-grant** [comment]
The vignette has been updated to show the correct package name.

**Comment by vjcitn** [comment]
Thank you.  You will need to look over the guidelines more carefully.  You have .RData files in several locations in the package.  Data for the package should be lodged in the /data folder.  The vignette has
```

> ## ----load library-------------------------------------------------------------
> library(seq.hotSPOT)
1/6 packages newly attached/loaded, see sessionInfo() for details.

> ## ----load data----------------------------------------------------------------
> load("example.RData")

> data
    chr          pos    gene
1     6 7.004926e+07  ADGRB3
2     6 7.004927e+07  ADGRB3
3     6 7.004926e+07  ADGRB3
```
and the dump goes on for many lines.  This is not appropriate in a vignette -- make a concise display, perhaps with `head()` or a tabulation of some sort.  On my 16GB laptop the process to run vignette code crashes every time.  Have you looked at memory consumption of your vignette code?  Does it really require more RAM to complete?

**Comment by sydney-grant** [comment]
I apologize for the miscommunication. Duplicates of the example data have been removed and now a single .RData file exists in a \data folder. The vignette has been updated to only display the first few lines of example data. Additionally, the code has been modified for less memory consumption. All functions in the vignette are now running smoothly on both my desktop and laptop, and this issue should be resolved.

**Comment by vjcitn** [comment]
The vignette has
```
load("..\\data\\example.RData")

```
This is not how packaging works.  `data(example)` would be used.  Please read Writing R Extensions with some care as you prepare your package.

**Comment by vjcitn** [comment]
If I correct some of the code, then
```
> data(example)

> head(data)
   chr       pos gene
22   4 126329603 FAT4
23   4 126329653 FAT4
24   4 126329608 FAT4
25   4 126329651 FAT4
26   4 126329633 FAT4
27   4 126329653 FAT4

> ## ----run amp finder-----------------------------------------------------------
> amps <- amp_pool(data, 100)
Error in amp_pool(data, 100) : argument "len" is missing, with no default
```

**Comment by sydney-grant** [comment]
Hello, thank you for your feedback. This issue has been corrected.

**Comment by vjcitn** [comment]
```
* checking examples ... ERROR
Running examples in 'seq.hotSPOT-Ex.R' failed
The error most likely occurred in:

> ### Name: amp_pool
> ### Title: create amplicon pool
> ### Aliases: amp_pool
> 
> ### ** Examples
> 
> 
> load(example.RData)
Error: object 'example.RData' not found
```
please run R CMD check.  please do not use "RData" suffix in the name of an object.  please do not use "example" to name an object, it interferes with the "example" function.

**Comment by sydney-grant** [comment]
This issue has been resolved. R CMD Check is running on my computer with no errors or warnings.

**Comment by sydney-grant** [comment]
Hello,

I am now subscribed to the bioc-devel mailing list. This error should be resolved.

Thank you,
Sydney

**Comment by lshep** [comment]
We don't see sydney.grant@roswellpark.org registered yet at https://stat.ethz.ch/pipermail/bioc-devel/ .  There is normally a confirmation email that gets sent that sometimes ends up in spam/junk by institutions, you might check to see if it was sent there.

**Comment by nilseling** [comment]
Hi @sydney-grant,

thanks for submitting `seq.hotSPOT` to Bioconductor. Please have a look at my comments below and once you have adjusted the package please provide a point-by-point response on how you addressed them.

## The DESCRIPTION file

- [x] It is not recommended to set `LazyData:` in the `DESCRIPTION` file to `true` but rather explicitly load the data via `data()`.
- [x] I would recommend to add https://github.com/sydney-grant/seq.hotSPOT to the `URL` field and https://github.com/sydney-grant/seq.hotSPOT/issues to the `BugReports` field.

## The NAMESPACE file

- [x] Please selectively import the functions that your package needs instead of the whole package. For example, instead of `@import hash` you would use `@importFrom hash make.keys`.

## The NEWS file

- [ ] Please add a `NEWS` file that keeps track on the changes you introduce to the package.
	  
## Package data

- [ ] Please have a look at https://r-pkgs.org/data.html#sec-documenting-data on how to document package data. The data man page should be accessible via `?mutation_data` and not `?data`.

## Documentation

- [x] I would recommend to use proper markdown formatting in the README file of your Github repository.

**Vignette**

- [ ] The vignette can be split into multiple sections (including a table of contents). One section should be labelled as "Introduction" and please also include an "Installation" section.
- [ ] In your Introduction section, please include a motivation for including your package in Bioconductor.
- [ ] It is recommended to use the `BiocStyle` package for vignette formatting.
- [x] There is no need to clear objects in the vignette (or the package functions) via (e.g.) `rm(list = ls()[! ls() %in% c("amps", "data", "fw_bins")])`.
- [x] Please add `sessionInfo()` to the end of your vignette.
- [x] I would recommend to explicitly call the function parameter in the vignette. For example: 

```
fw_bins <- fw_hotspot(bins = amps, 
                      data = data, 
                      amp = 100, 
                      len = 1000)
```

- [x] Please double check the formatting of the lists. It seems like the sub-lists (a., b., etc.) are not correctly formatted. 

**Man pages**

- [x] Please provide a package man page that can be accessed via `?seq.hotSPOT`.

## Unit tests

- [x] Please consider adding unit tests that test the main functions of the package. This will facilitate future maintenance of the package.

## R code

- [ ] Please address the following issues:

```
amp_pool: no visible global function definition for ‘ftable’
amp_pool: no visible binding for global variable ‘chr’
com_hotspot: no visible global function definition for ‘ftable’
com_hotspot: no visible binding for global variable ‘id’
com_hotspot: no visible global function definition for ‘median’
fw_hotspot: no visible global function definition for ‘ftable’
fw_hotspot: no visible binding for global variable ‘id’
fw_hotspot: no visible binding for global variable ‘count’
Undefined global functions or variables:
  chr count ftable id median
Consider adding
  importFrom("stats", "ftable", "median")
to your NAMESPACE file.
```

- [ ] It is also recommended to address the following:

```
Avoid 1:...; use seq_len() or seq_along()
Avoid using '=' for assignment and use '<-' instead
```

- [x] The function `amplicon_finder` is defined within the `amp_pool` function. I would recommend to define the helper function `.amplicon_finder` outside of the `amp_pool` function and only call it from within the `amp_pool` function.
- [x] Please write validity checks for function inputs. For example, you want to test, if the `data` data.frame contains all relevant columns. Please test similar things for all inputs.
- [x] You are using quite a few `for` loops which could potentially be replaced by vectorized operations. While it's not mandatory to change this at the moment it might increase the performance and scalability of your code.

**Comment by sydney-grant** [comment]
> Hi @nilseling,
> 
> Thank you for taking the time to review my packages. All of your comments and suggestions are much appreciated. I have listed below the tasks I have completed an explanation on how I have addressed each comment. Please let me know if there are further changes that need to be made.
> 
> ## The DESCRIPTION file
> * [x]  It is not recommended to set `LazyData:` in the `DESCRIPTION` file to `true` but rather explicitly load the data via `data()`.
> **LazyData in description file has been set to FALSE. Data is loaded via data().**
> * [x]  I would recommend to add https://github.com/sydney-grant/seq.hotSPOT to the `URL` field and https://github.com/sydney-grant/seq.hotSPOT/issues to the `BugReports` field.
> **These links have been added to a URL and BugReports section of the description file.**
> ## The NAMESPACE file
> * [x]  Please selectively import the functions that your package needs instead of the whole package. For example, instead of `@import hash` you would use `@importFrom hash make.keys`.
>  **This change has been made in all .R files. Only specific functions used in the package are now loaded**
> ## The NEWS file
> * [x]  Please add a `NEWS` file that keeps track on the changes you introduce to the package.
>  **A NEWS file has been created.**
> ## Package data
> * [x]  Please have a look at https://r-pkgs.org/data.html#sec-documenting-data on how to document package data. The data man page should be accessible via `?mutation_data` and not `?data`.
>  **I have reviewed this webpage and the documentation of example data has been updated accordingly.**
> ## Documentation
> * [x]  I would recommend to use proper markdown formatting in the README file of your Github repository.
>  **README file has been updated for proper formatting.**
> **Vignette**
> 
> * [x]  The vignette can be split into multiple sections (including a table of contents). One section should be labelled as "Introduction" and please also include an "Installation" section.
> * [x]  In your Introduction section, please include a motivation for including your package in Bioconductor.
> * [x]  It is recommended to use the `BiocStyle` package for vignette formatting.
> * [x]  There is no need to clear objects in the vignette (or the package functions) via (e.g.) `rm(list = ls()[! ls() %in% c("amps", "data", "fw_bins")])`.
> * [x]  Please add `sessionInfo()` to the end of your vignette.
> * [x]  I would recommend to explicitly call the function parameter in the vignette. For example:
> 
> ```
> fw_bins <- fw_hotspot(bins = amps, 
>                       data = data, 
>                       amp = 100, 
>                       len = 1000)
> ```
> 
> * [x]  Please double check the formatting of the lists. It seems like the sub-lists (a., b., etc.) are not correctly formatted.
> **Table of contents including introduction and installation has been added to vignette. Introduction has been added describing importance and relevance of package for NGS experiments. Formatting has been changed to BiocStyle, and remaining modification have been made.**
> **Man pages**
> 
> * [x]  Please provide a package man page that can be accessed via `?seq.hotSPOT`.
> **A man page for seq.hotSPOT has been added.**
> ## Unit tests
> * [x]  Please consider adding unit tests that test the main functions of the package. This will facilitate future maintenance of the package.
> **A unit test has been added for the functions fw_hotspot and com_hotspot. This test double checks the measured output of the functions and ensures mutations are not being counted more then one time.**
> ## R code
> * [x]  Please address the following issues:
> 
> ```
> amp_pool: no visible global function definition for ‘ftable’
> amp_pool: no visible binding for global variable ‘chr’
> com_hotspot: no visible global function definition for ‘ftable’
> com_hotspot: no visible binding for global variable ‘id’
> com_hotspot: no visible global function definition for ‘median’
> fw_hotspot: no visible global function definition for ‘ftable’
> fw_hotspot: no visible binding for global variable ‘id’
> fw_hotspot: no visible binding for global variable ‘count’
> Undefined global functions or variables:
>   chr count ftable id median
> Consider adding
>   importFrom("stats", "ftable", "median")
> to your NAMESPACE file.
> ```
> **I have added imports for stats ftable and median. However, I do not believe the variables chr, count, and id should be defined as global variables for this package.**
> * [x]  It is also recommended to address the following:
> 
> ```
> Avoid 1:...; use seq_len() or seq_along()
> Avoid using '=' for assignment and use '<-' instead
> ```
> **These changes have been made in the appropriate places within the code of each function.**
> * [x]  The function `amplicon_finder` is defined within the `amp_pool` function. I would recommend to define the helper function `.amplicon_finder` outside of the `amp_pool` function and only call it from within the `amp_pool` function.
> **The function amplicon_finder has been defined as an internal function and is called within the amp_pool function.**
> * [x]  Please write validity checks for function inputs. For example, you want to test, if the `data` data.frame contains all relevant columns. Please test similar things for all inputs.
> **Validity checks have been added for all function input variables.**
> * [x]  You are using quite a few `for` loops which could potentially be replaced by vectorized operations. While it's not mandatory to change this at the moment it might increase the performance and scalability of your code.
> **At this time, I do not wish to substitute for loops for vectorized operations.**

**Comment by nilseling** [comment]
Hi @sydney-grant 

thanks for addressing the comments! I noticed a couple of minor things that should be addressed before acceptance:

- [ ] The NEWS file is not correctly formatted. Please have a look at: https://contributions.bioconductor.org/news.html#news
- [ ] Please change `@aliases data` to `@aliases mutation_data` in the docs. The call `?data` still refers to the `seq.hotSPOT` documentation.
- [ ] Please add an "Installation" section to the vignette describing the Bioconductor installation instructions.
- [ ] In the vignette please add a motivation for inclusion into Bioconductor. Does Bioconductor provide similar tools or tools that can be used to perform pre-/post-processing of your data?
- [ ] You could use `globalVariables()` to specify `chr`, `count`, `id` as global variables. Have a look at https://www.r-bloggers.com/2019/08/no-visible-binding-for-global-variable/ for alternative approaches.

**Comment by sydney-grant** [comment]
> Hi @nilseling 
> 
> Thank you for your quick feedback. Please see my comments below:
> 
> * [x]  The NEWS file is not correctly formatted. Please have a look at: https://contributions.bioconductor.org/news.html#news
> **The NEWS file formatting has been switched to .md formatting style.**
> * [x]  Please change `@aliases data` to `@aliases mutation_data` in the docs. The call `?data` still refers to the `seq.hotSPOT` documentation.
> **This correction has been made.**
> * [x]  Please add an "Installation" section to the vignette describing the Bioconductor installation instructions.
> **Installation section has been added to vignette.**
> * [x]  In the vignette please add a motivation for inclusion into Bioconductor. Does Bioconductor provide similar tools or tools that can be used to perform pre-/post-processing of your data?
> **Additional text has been added to the introduction explaining the gap in available tools similar to seq.hotSPOT and how RTCGA Bioconductor packages can be used to extract data to use for seq.hotSPOT.**
> * [x]  You could use `globalVariables()` to specify `chr`, `count`, `id` as global variables. Have a look at https://www.r-bloggers.com/2019/08/no-visible-binding-for-global-variable/ for alternative approaches.
**"chr", "count", and "id" have been added as global variables.**

**Comment by nilseling** [comment]
Great, thanks! Please fix the formatting of the "Installation" call. You should provide the installation instructions in an `eval=FALSE` R code chunk. Once this is done I can accept the package.

**Comment by sydney-grant** [comment]
@nilseling I believe this should now be corrected.

**Comment by nilseling** [comment]
Thanks! This looks good and I can accept the package.

**Comment by lshep** [comment]
The default branch of your GitHub repository has been added to Bioconductor's
git repository as branch devel.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/sydney-grant.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("seq.hotSPOT")`. The package 'landing page' will be created at

https://bioconductor.org/packages/seq.hotSPOT

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
